### PR TITLE
[KOGITO-3034][BXMSPROD-1094] Setup Jenkins DSL scripts

### PR DESCRIPTION
**Pull request by radtriste**
https://issues.redhat.com/browse/KOGITO-3034
https://issues.redhat.com/browse/BXMSPROD-1094

Related PRs:
- https://github.com/kiegroup/kogito-pipelines/pull/92
- https://github.com/kiegroup/kogito-runtimes/pull/940
- https://github.com/kiegroup/kogito-apps/pull/581
- https://github.com/kiegroup/optaplanner/pull/1085
- https://github.com/kiegroup/kogito-examples/pull/500
- https://github.com/kiegroup/kogito-images/pull/307
- https://github.com/kiegroup/kogito-cloud-operator/pull/710

**Setup DSL scripts for the different Kogito jobs.**

- The repository `kogito-pipelines` does contain the seed job and some configuration to generate the different jobs.
- Each Kogito repository keeps the definition of the different jobs. It will contain just some groovy files + a test script which will checkout `kogito-pipelines` for testing the different script generation.
- Seed job will be configured with different branch and will so checkout those branches for job creation. 
Configuration can be found in `dsl/seed/config.yaml`. 
Example of configuration:
```yaml
# The configuration present here will be available as uppercase env vars for job generation

jenkins_email_creds_id: KOGITO_CI_EMAIL_TO

git:
  # Each branch will be accessible as `GIT_BRANCH` env var
  branches:
  - master
  - 1.0.x
  author:
    name: kiegroup
    credentials_id: kie-ci
    token_credentials_id: kie-ci2-token
  bot_author:
    name: bsig-gh-bot
    credentials_id: bsig-gh-bot
  # Each main branch will be accessible as `GIT_MAIN_BRANCH` env var
  main_branch:
    default: master

maven:
  settings_file_id: kogito_release_settings
  nexus:
    release_url: https://repository.jboss.org/nexus
    release_repository: jboss-releases-repository
    staging_profile_url: https://repository.jboss.org/nexus/content/groups/kogito-public/
    staging_profile_id: 2161b7b8da0080
    build_promotion_profile_id: ea49ccd6f174
  ## artifacts_repository Can be set if needed a different deploy repository than usual. Mostly for test purpose.
  artifacts_repository: ''

cloud:
  image:
    registry_credentials_nightly: nightly_kogito
    registry_credentials_release: release_kogito
    registry: 'quay.io'
    namespace: 'kiegroup'
    name_suffix: 'nightly'
    latest_git_branch: master
```
**NOTE: Only the `git.branches` should change over time, depending on the different release branch we want to manage.**
_This configuration is then available as env variables for job generation._
- Each branch will contain a configuration of repositories to checkout, in `dsl/branch_config.yaml`
Simple example for 1.0.x branch:
```
dependent_repositories:
  - name: kogito-runtimes
  - name: kogito-apps
  - name: optaplanner
    branch: 8.0.x
  - name: kogito-examples
  - name: kogito-images
  - name: kogito-cloud-operator
```
_NOTE: You could also redefine `author` and `bot_author` information if needed_

More information can also be found into the README file

### CODE REVIEW

**Comment by radtriste** [comment]
This is a big one and important for the automation of all jenkins jobs in Kogito.
Once that one is merged, I will set all related PRs as ready for review.
Once all is merged, we can start to generate the jobs and migrate to full automation on Jenkins (will be done after 1.1.0 release of course)

**Comment by radtriste** [comment]
FYI, all generated jobs can be found here: https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/custom/job/tradisso/job/dsl_test/

File: dsl/seed/Jenkinsfile.seed [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+// Map<branch, Map<repo, config>>
+branchConfigs = [:]
+
+PIPELINES_REPO = 'kogito-pipelines'
+
+// Configuration of the pipeline is done via the `config.yaml` file
+pipeline {
+    agent {
+        label 'kie-rhel7 && kie-mem4g'
+        // label 'kieci-02-docker'
```
Comment (Ginxo) [inline dsl/seed/Jenkinsfile.seed]: to be removed?

**Review by Ginxo — APPROVED** [review]
From technical POV is perfect. I can't tell from business POV.

File: dsl/seed/Jenkinsfile.seed (hunk 1 above)
Comment (radtriste) [reply dsl/seed/Jenkinsfile.seed]: done

**Review by mbiarnes — APPROVED** [review]
+1

**Review by msmagnanijr — APPROVED** [review]
lgtm

**Review by Kaitou786 — APPROVED** [review]

**Comment by radtriste** [comment]
In latest commits, I just simplified the model of creating jobs by just using a `jobs.groovy` so we can easily reuse some methods.
This was done also to easily create jobs introduced by the `runtimes bdd tests`.
I guess in the future, we could think of a registration model of jobs before generation on a specific branch, but that is far more complex to setup IMO and I think we can start with this model for now

File: Jenkinsfile.create-release-branches [hunk 2]
```diff
@@ -6,18 +6,13 @@ pipeline {
...
     }
 
-    parameters {
-        string(name: 'REPOSITORIES', defaultValue: '', description: 'Comma-separated list of repository[:base branch] to update. The default base branch for every repository is the \'master\' branch.')
-        string(name: 'RELEASE_BRANCH', defaultValue: '', description: 'Release branch to create')
-        
-        string(name: 'GIT_AUTHOR', defaultValue: 'kiegroup', description: 'Which Git author repository ?')
-        string(name: 'GIT_AUTHOR_CREDS_ID', defaultValue: 'kie-ci', description: 'Git author credentials id')
+    // parameters {
+    // For parameters, check into .jenkins/dsl/jobs/nightly_release.groovy file
```
Comment (Kevin-Mok) [inline Jenkinsfile.create-release-branches]:
```suggestion
    // For parameters, check into .jenkins/dsl/jobs.groovy file
```

File: Jenkinsfile.nightly [hunk 3]
```diff
@@ -18,61 +19,36 @@ OPERATOR_PROMOTE = 'kogito-operator-promote'
...
     options {
-        buildDiscarder logRotator(artifactDaysToKeepStr: '', artifactNumToKeepStr: '', daysToKeepStr: '', numToKeepStr: '30')
         timeout(time: 720, unit: 'MINUTES')
     }
 
-    parameters {
-        string(name: 'BUILD_BRANCH_NAME', defaultValue: '', description: 'Which branch to build? Set if you are not on a multibranch pipeline.')
-        string(name: 'GIT_AUTHOR', defaultValue: 'kiegroup', description: 'Which Git author repository ?')
+    // parameters {
+    // For parameters, check into .jenkins/dsl/jobs/nightly_release.groovy file
```
Comment (Kevin-Mok) [inline Jenkinsfile.nightly]:
```suggestion
    // For parameters, check into .jenkins/dsl/jobs.groovy file
```

File: Jenkinsfile.release [hunk 4]
```diff
@@ -31,62 +31,18 @@ pipeline {
...
-        booleanParam(name: 'SKIP_IMAGES_PROMOTE', defaultValue: false, description: 'To skip Images Promote only. Automatically skipped if SKIP_IMAGES_DEPLOY is true')
-        booleanParam(name: 'SKIP_OPERATOR_DEPLOY', defaultValue: false, description: 'To skip Operator Test & Deployment.')
-        booleanParam(name: 'SKIP_OPERATOR_PROMOTE', defaultValue: false, description: 'To skip Operator Promote only. Automatically skipped if SKIP_OPERATOR_DEPLOY is true.')
-
-        booleanParam(name: 'USE_TEMP_OPENSHIFT_REGISTRY', defaultValue: false, description: 'If enabled, use Openshift registry to push temporary images')
-
-
-    }
+    // parameters {
+    // For parameters, check into .jenkins/dsl/jobs/nightly_release.groovy file
```
Comment (Kevin-Mok) [inline Jenkinsfile.release]:
```suggestion
    // For parameters, check into .jenkins/dsl/jobs.groovy file
```

File: dsl/seed/jobs/seed_job.groovy [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// +++++++++++++++++++++++++++++++++++++++++++ create a seed job ++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+// Configuration of the seed and generated jobs is done via `jenkins/dsl/seed/config.properties`
```
Comment (Kevin-Mok) [inline dsl/seed/jobs/seed_job.groovy]:
```suggestion
// Configuration of the seed and generated jobs is done via `dsl/seed/config.yaml`
```

File: docs/jenkins.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+This configuration contains the configuration for a specific branch.
+
+By default, there is a list of dependent git repositories where job scripts will be taken for jobs generation.  
+For each of them, you can override the global Git information (author name, credentials ...).
+
+### Seed job testing
+
+    ```bash
+    $ cd dsl/seed && ./gradlew test
+    ```
```
Comment (Kevin-Mok) [inline docs/jenkins.md]:
```suggestion
```bash
$ cd dsl/seed && ./gradlew test
```

The 4 spaces are making the entire thing show up as a code block (see [resulting page](https://github.com/radtriste/kogito-pipelines/blob/setup_dsl/docs/jenkins.md#seed-job-testing)).

File: docs/jenkins.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+The script clones the `kogito-pipelines` repository and then call the `seed_test.sh` which should copy the jobs and 
+
+Then you can call the script:
+
+    ```bash
+    $ chmod u+x .jenkins/dsl/test.sh
+    $ cd .jenkins/dsl && ./scripts/test.sh
+    ```
```
Comment (Kevin-Mok) [inline docs/jenkins.md]: Spaces should be removed here as well.

File: dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/templates/KogitoJobTemplate.groovy [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                            addTestResults(true)
+                            showMatrixStatus(false)
+                            statusUrl('${BUILD_URL}display/redirect')
+                            triggeredStatus('Build triggered.')
+                            startedStatus('Build started.')
+                        }
+                        ghprbBuildStatus {
+                            messages {
+                                /* need to manually add the "$" before the bracketed BUILD_URL
+                                in the final so that the seed build URL is not used */
```
Comment (Kevin-Mok) [inline dsl/seed/src/main/groovy/org/kie/jenkins/jobdsl/templates/KogitoJobTemplate.groovy]:
```suggestion
```

**Review by Kevin-Mok — CHANGES_REQUESTED** [review]

**Comment by radtriste** [comment]
@Kevin-Mok updated with your suggestions

**Review by Kevin-Mok — APPROVED** [review]

**Pull request merged**